### PR TITLE
[kernels][fused_moe] Add another fused EP MoE kernels

### PR DESCRIPTION
**Pull request by rupengliu-meta**
Summary:
Upload a new kernel by @rupengliu-meta @bangshengtang @zongweiz @yaochengji  . Bring the fused expert-parallel MoE TPU kernels into the repo as a self-contained, dependency-free package. The main kernel runs the entire EP MoE (gather, GMM1, activation, GMM2) with the ICI A2A fused into a single Pallas call, replacing the usual post-kernel all-reduce and keeping the collective overlapped with compute. A second, smaller kernel is included as a worked example of fusing the upstream all-gather into the grouped matmul (per-round push schedule) and is not performance-tuned.

The code is fully standalone: device-specific tuned block-size tables and all internal references were stripped for external release, the two kernels share a single trimmed grouped-matmul base, and a README documents usage and future optimization directions.

Perf gain:
Currently we see 15-55% perf gain for entire MoE layer compared with existing MoE implementation; Around 10%-30% e2e gain on different batch_sizes/prefill_decode mode with this kernel

# Description

Start with a short description of what the PR does and how this is a change from
the past.

The rest of the description includes relevant details and context, examples:
- why is this change being made,
- the problem being solved and any relevant context,
- why this is a good solution,
- some information about the specific implementation,
- shortcomings of the solution and possible future improvements.

If the change fixes a Github issue, please include a link, e.g.,:
FIXES: #123456

# Tests

Please describe how you tested this change, and include any instructions and/or
commands to reproduce.

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

**Comment by helloworld1** [comment]
@rupengliu-meta Thanks a lot for the contribution! Do you have a branch to integrate the kernel with the model like Qwen 3.5 or Qwen coder so we can compare the performance E2E? Thank you!

**Comment by rupengliu-meta** [comment]
@helloworld1 @kyuyeunk I will try my best to find the bandwidth in the following two weeks or so to do the integration. Also I already added an example integration file. We have an internal benchmarking dashboard with inhouse model, as mentioned, the current fused moe kernel outperforms the existing moe implementation by 15-50% for moe layer and ~10-25% e2e, the gain is more with larger batch size

**Comment by helloworld1** [comment]
I did a hack integration https://github.com/vllm-project/tpu-inference/commit/08b03477ecf3a65a6f6b69dc757b804244e03ad3 and benchmark the qwen 3.5 397B model on 8K/1K concurrency 256 case.

MoE layer latency
Prefill
gmm_v2: 1.9ms
gmm v2 rs kernel: 2.1ms

Decode:
gmm_v2: 370us
gmm v2 rs kernel: 330us

**Comment by rupengliu-meta** [comment]
> I did a hack integration [08b0347](https://github.com/vllm-project/tpu-inference/commit/08b03477ecf3a65a6f6b69dc757b804244e03ad3) and benchmark the qwen 3.5 397B model on 8K/1K concurrency 256 case.
> 
> MoE layer latency Prefill gmm_v2: 1.9ms gmm v2 rs kernel: 2.1ms
> 
> Decode: gmm_v2: 370us gmm v2 rs kernel: 330us

what is the block sizes you are using, the block sizes need to trigger the expert weight caching, which avoid weights loading during repeated gmm_tile, feel free to share the kernel trace with your benchmark, prefill should see the bigger gain.

**Comment by rupengliu-meta** [comment]
ok, from your PR, it seems you didn't do any block size tuning?

**Comment by helloworld1** [comment]
@rupengliu-meta Could you give us some insight / heuristic for a good tuning?

**Comment by rupengliu-meta** [comment]
> @rupengliu-meta Could you give us some insight / heuristic for a good tuning?

yes, please use below heuristic:

same-expert weight caching. It's not a flag; it turns on automatically when the buffers can hold all of one expert's weight tiles at once:

num_w1_bufs >= (n1/tile_n1) * (k1/tile_k1)
num_w2_bufs >= (n2/tile_n2) * (k2/tile_k2)
When that holds and consecutive token-tiles hit the same expert, the kernel skips reloading weights (no re-DMA, no wait). So the tradeoff is: coarser tiles + more buffers = caching, but higher VMEM.

Core heuristic (split by M):

Small M (each expert touched ~once): caching is useless. Use full-N tiles, single buffer (…,1,1) to minimize per-step overhead.
Large M (each expert spans many token-tiles): caching wins big (kills ~90% of weight DMAs). Use smaller-N tiles + (…,2,2) or (…,3,3) so bufs ≥ steps.
Very large M: bump tile_m (e.g. 128→256) so more compute per tile hides the reduce-scatter latency.

FP8: at decode, you should see hbm bandwidth ultilization reaches 80%+, mostly bandwidth bound. At prefill, it should be purely compute bound, you shouldn't see any dma wait and a2a collective latency etc

The kernel supports up to 16k max batched token count right now

**Review by QiliangCui — APPROVED** [review]

File: tpu_inference/kernels/experimental/fused_moe/gmm_fused_rs_nodedup.py [hunk 1]
```diff
@@ -0,0 +1,2044 @@
...
+                    src_ref=w1_scale_ref.at[
+                        expert_id,
+                        pl.ds(k_id * nqb1, nqb1),
+                        :,
+                        pl.ds(n_id * w1_dma_n, w1_dma_n),
+                    ],
+                    dst_ref=w1_scale_buf_ref.at[buf_id],
+                    sem=w1_sem_ref.at[buf_id],
+                ).start()
+        if has_bias:
```
Comment (qizzzh) [inline tpu_inference/kernels/experimental/fused_moe/gmm_fused_rs_nodedup.py:585]: The weight DMA at 519–549 and scale DMA at 550–584 branch on `act_fn and num_n1 > 1` to load gate half from `_gate_offset = n_id*tile_n1` and up half from `_up_offset = out_n1 + n_id*tile_n1`. The bias DMA here is unconditional and loads a contiguous `2*tile_n1` slab from `pl.ds(n_id * w1_dma_n, w1_dma_n)`. Compute at 673–687 then slices the buffer as `[gate | up]`.

**Failure:** with `act_fn='silu'`, `has_bias=True`, and `intermediate_size` large enough that `num_n1 > 1` (e.g. `intermediate_size=1024`, `tile_n1=256`, `num_n1=4`), `w1_bias_buf_ref[:, tile_n1:2*tile_n1]` contains the *next* gate slice instead of the up bias (which lives at `w1_bias[E, :, out_n1 + n_id*tile_n1 : ...]`). GMM1 adds gate biases where up biases belong — silent wrong-answer for any MoE-with-MLP-bias model (GPT-OSS-style).

Note: `w2_bias` DMA at 615–616 is fine — GMM2 has no fused activation and no gate/up split.

**Fix:** mirror the weight/scale DMA branching: when `act_fn and num_n1 > 1`, issue two copies for the gate and up halves into `w1_bias_buf_ref.at[:, :tile_n1]` and `w1_bias_buf_ref.at[:, tile_n1:]`.

File: tpu_inference/kernels/experimental/fused_moe/fused_moe_rs.py [hunk 2]
```diff
@@ -0,0 +1,528 @@
...
+            valid_mask = (col_idx < norm_size)[None, None, :]
+            out_f32 = out_3d.astype(jnp.float32) * valid_mask
+            pnw = jnp.concatenate([
+                pnw_raw,
+                jnp.zeros(hidden_size - norm_size, dtype=jnp.float32)
+            ])
+        else:
+            out_f32 = out_3d.astype(jnp.float32)
+            pnw = pnw_raw
+        out_f32 = out_f32 * pnw[None, None, :]
```
Comment (qizzzh) [inline tpu_inference/kernels/experimental/fused_moe/fused_moe_rs.py:323]: Lines 323–325 compute:

```
out_f32 = out_f32 * pnw
var     = mean(out_f32 ** 2)
out_3d  = out_f32 * rsqrt(var + 1e-8)
```

i.e. `y = (x*w) * rsqrt(mean((x*w)^2) + eps)`. Standard RMSNorm (cf. `tpu_inference/layers/jax/layers.py:96–98`) is `var = mean(x^2)`, `normed = x * rsqrt(var + eps)`, then `normed *= scale` — normalize FIRST, scale LAST.

**Failure:** for `x=[1,1]`, `w=[0,1]` (so `pnw = w + 1 = [1,2]`):
- reference: `rms(x)=1`, `normed=[1,1]`, scaled `= [1.0, 2.0]`
- kernel:    `out_f32=[1,2]`, `var=(1+4)/2=2.5`, `rsqrt≈0.6325`, output `≈[0.6325, 1.265]` (>36% divergence)

For uniform `w=1` the output is **half** the reference — the weight is normalized away.

**Fix:** compute variance from `out_f32` before applying `pnw`, then multiply the normalized result by `pnw`.

File: tpu_inference/kernels/experimental/fused_moe/fused_moe_rs.py [hunk 3]
```diff
@@ -0,0 +1,528 @@
...
+    expert_axis = get_moe_expert_axis(mesh, EXPERT)
+    ep_size = get_mesh_shape_product(mesh, expert_axis)
+    ep_p_spec = P(expert_axis)
+    data_p_spec = P(MLP_DATA)
+    # SP off: hidden replicated in/out with an explicit all-gather in the body.
+    sp_enabled = enabled_tpu_sp()
+    hidden_in_spec = data_p_spec if sp_enabled else P()
+    moe_out_spec = (P(combine_partition_axes(MLP_DATA, expert_axis))
+                    if sp_enabled else P())
+    fp8_post_gather = ((not sp_enabled) and w1_scale is not None
```
Comment (qizzzh) [inline tpu_inference/kernels/experimental/fused_moe/fused_moe_rs.py:369]: The reassignment overrides the caller's `True` if `sp_enabled` OR `w1_scale is None` OR `w2_scale is None`. The flag has double duty: it drives the kernel's `fp8_direct_write` (which does require quantized weights)
  AND `_all_gather_token_hidden`'s activation quant (which does not depend on weight quant). So on SP-off + bf16 the caller opting into comm compression gets silently downgraded to bf16 with no log.

  **Fix:** either split into two flags (`kernel_fp8_direct_write` vs `comm_fp8_activation`), or log at INFO when a caller-supplied `True` is downgraded.

**Review by qizzzh — APPROVED** [review]

**Comment by rupengliu-meta** [comment]
no need to merge the code, feel free to optimize it internally, we can share the optimized code offline. much appreciated! (please optimize the fp4 code path, thanks)

**Comment by helloworld1** [comment]
I did the micro benchmark and found the kernel in general performs better for higher number of tokens to a certain point. I tried different combo of tiling combos and heuristic without finding a config that can work well across mulitple workloads. I think selectively using this kernel for based on the token size and workload can provide E2E performance gain. 

                                                       
| Xprof Test Name | `jit(fused_moe_func)` Median Time | `jit(fused_moe_func_rs)` Median Time | `_rs` Better? |
| :--- | :--- | :--- | :--- |
| `qwen35_tokens_8` | 103.16 us (0.1032 ms) | 124.66 us (0.1247 ms) | **No** (20.8% slower) |
| `qwen35_tokens_16` | 137.51 us (0.1375 ms) | 172.47 us (0.1725 ms) | **No** (25.4% slower) |
| `qwen35_tokens_32` | 198.03 us (0.1980 ms) | 260.41 us (0.2604 ms) | **No** (31.5% slower) |
| `qwen35_tokens_64` | 304.08 us (0.3041 ms) | 332.89 us (0.3329 ms) | **No** (9.5% slower) |
| `qwen35_tokens_128` | 372.13 us (0.3721 ms) | 375.31 us (0.3753 ms) | **No** (0.9% slower) |
| `qwen35_tokens_256` | 415.02 us (0.4150 ms) | 405.81 us (0.4058 ms) | **Yes** (2.2% faster) |
| `qwen35_tokens_512` | 468.39 us (0.4684 ms) | 426.13 us (0.4261 ms) | **Yes** (9.0% faster) |
| `qwen35_tokens_1024` | 665.64 us (0.6656 ms) | 458.67 us (0.4587 ms) | **Yes** (31.1% faster) |
| `qwen35_tokens_2048` | 1491.68 us (1.4917 ms) | 573.02 us (0.5730 ms) | **Yes** (61.6% faster) |
| `qwen35_tokens_4096` | 1001.22 us (1.0012 ms) | 1039.91 us (1.0399 ms) | **No** (3.9% slower) |
| `qwen35_tokens_8192` | 1317.12 us (1.3171 ms) | 65429.28 us (65.4293 ms) | **No** (4867.6% slower) |
| `qwen_coder_8` | 197.64 us (0.1976 ms) | 207.93 us (0.2079 ms) | **No** (5.2% slower) |
| `qwen_coder_16` | 293.75 us (0.2937 ms) | 325.82 us (0.3258 ms) | **No** (10.9% slower) |
| `qwen_coder_32` | 331.44 us (0.3314 ms) | 377.29 us (0.3773 ms) | **No** (13.8% slower) |
| `qwen_coder_64` | 374.90 us (0.3749 ms) | 418.75 us (0.4187 ms) | **No** (11.7% slower) |
| `qwen_coder_128` | 397.71 us (0.3977 ms) | 429.23 us (0.4292 ms) | **No** (7.9% slower) |
| `qwen_coder_256` | 428.61 us (0.4286 ms) | 439.47 us (0.4395 ms) | **No** (2.5% slower) |
| `qwen_coder_512` | 533.30 us (0.5333 ms) | 459.80 us (0.4598 ms) | **Yes** (13.8% faster) |
| `qwen_coder_1024` | 854.14 us (0.8541 ms) | 620.53 us (0.6205 ms) | **Yes** (27.4% faster) |
| `qwen_coder_2048` | 1959.33 us (1.9593 ms) | 1016.04 us (1.0160 ms) | **Yes** (48.1% faster) |
| `qwen_coder_4096` | 6242.75 us (6.2428 ms) | 1894.44 us (1.8944 ms) | **Yes** (69.7% faster) |
| `qwen_coder_8192` | 1039.32 us (1.0393 ms) | 3511.35 us (3.5113 ms) | **No** (237.9% slower) |

**Review by helloworld1 — APPROVED** [review]
The kernel is good starting point as mega fused kernel. Though multiple optimization / improvement can be made here. The block size tuning logic can be improved. The util gmm_v2_gather_scatter has many code sharable with fused_moe_gmm.  But for now it is good to keep everything encapsulated in the fused_moe directory.

I would change to "fused_moe_v2" to avoid confusion with existing fused_moe kernel.

**Comment by rupengliu-meta** [comment]
qwen35_tokens_8 | 103.16 us (0.1032 ms) | 124.66 us (0.1247 ms) | No (20.8% slower)
-- | -- | -- | --
qwen35_tokens_16 | 137.51 us (0.1375 ms) | 172.47 us (0.1725 ms) | No (25.4% slower)
qwen35_tokens_32 | 198.03 us (0.1980 ms) | 260.41 us (0.2604 ms) | No (31.5% slower)
qwen35_tokens_64 | 304.08 us (0.3041 ms) | 332.89 us (0.3329 ms) | No (9.5% slower)

this area is slower is expected due to the tiling restriction of the moe kernel. And this domain is not meaningful for us at all (with speculative decoding)

**Review by kyuyeunk — APPROVED** [review]
